### DISCUSSION TITLE
Appointments (Malawi Core)

### DISCUSSION THREAD

**Discussion by roywanyaga** [opensrp/fhircore#1954] [Ideas]
# Context

**Appointments:** Where all persons who have been booked for a clinic appointment are listed and the clinic summary can be viewed. An appointment is at a date level, not specific time.

In Malawi Core Patients are added to the appointment list when a clinic appointment date is set in the following questionnaires:

 1. **Finish Visit Questionnaire:** Setting a date of the next appointment in the Finish Visit questionnaire will add the client on the appointments list on the day that has been entered.
 2. **Tracing Outcome Questionnaires:** Setting a date of clinic appointment in the tracing outcomes questionnaire will add a record of the client to the appointments list.

- The client will appear twice on the appointment list if they have been added to the appointment list through both of the questionnaires; one for each. 

-   Entering a new date of next appointment in the _**Finish Visit questionnaire**_ before or on the day of the clients appointment is due creates a new listing for the client in which they no longer appear on the date for ‘today’ or the date they were supposed to but a new date in the future
    
-   Entering a new date of appointment in the _**Finish Visit questionnaire**_ after the day the client's appointment was due (i.e. the client missed their appointment, creates a new listing for the client in the future but leaves the missed appointment. This way, if anyone looks at past dates, the people listed there will be those that missed their appointment only.

The reasons for addition of patients to the appointment list are based on the following three broad patient categories within Malawi Core: 

 1. **ART Clients**
 2. **Exposed Infants**
 3. **Contacts and Community Positives:** including a) **`Child Contacts`**, b) **`Persons who is Reactive at the Community`**, and c) **`Sexual Contacts`**

The **ART Clients** will be on the Appointments List for 7 reasons:
 1.  **Cervical Cancer Screening** – This reason will be added if a *non-pregnant woman* of *25 years+* *has not had a cervical cancer screen* for *a year or more* based on the *date of the last cervical cancer screen* set in the **`Women’s Health Screening Questionnaire`**
2.  **Index Case Testing** – This reason will be added whenever *viral load needs collection* (the **`viral load collection questionnaire`** is *activated*) or when a client has a high *viral load result* entered in the **`Viral Load Results questionnaire`**    
3.  **Linkage** – This reason is added when the *date of clinic appointment* is added through the **`Tracing Outcome questionnaire`** for an ART Client with the reason ‘*Linkage*’ is entered.    
4.  **Refill** – This reason is the **`default reason`** set for *every ART client* and is replaced with the other 6 listed reasons    
5.  **Viral Load Collection** – This reason is added when the *results of a viral load* are *missing* or *invalid* as stated in the **`Viral Load Results Questionnaire`**. This reason is also added if  *the date of the last viral load collected is 6 months ago for pregnant and breastfeeding women* (based on the **`vitals and viral load collection questionnaires`**) or on *a child regimen* (as based on the **`TB History and Regimen and viral load collection questionnaire`**); *1 year ago* for *non-pregnant people on adult regimen* (based on the **`vitals and viral load collection questionnaires`**); *3 months ago* from the *High Viral Load Welcome Service* provided in the **`Welcome Service Questionnaire`**; *6 months since regimen change* (based on the **`TB History and Regimen questionnaire`**)    
6.  **Welcome Service** – This reason is added if an *ART client misses the appointment* set in the **`FINISH VISIT questionnaire`** by *28 days*; If they have a *High Viral Load result* entered in the **`Viral Load Results questionnaire`** or are an *ART client in the first 6 months of Care*    
7.  **Welcome Service Follow Up** – This reason is added if an ART client states that they would like a *Welcome Service follow up visit* conducted in the **`Welcome Service`** or the **`Welcome Service Follow up Questionnaires`**   
**NB:** *An ART client can have more than one reason listed on the appointments list. All reasons listed replace the reason ‘Refill’*

The **Exposed Infant** will be on the tracing list for 3 reasons:
1.  **Dry Blood Sample Positive** – This reason will be added if an exposed infant had a *positive HIV test result* entered in the **`HIV test and Results Questionnaire`**    
2.  **Milestone HIV Test** - This reason will be added if the exposed infant is *6 weeks, 1 year, 2years old* or is *older than that but has not had a milestone HIV test.*    
3.  **Routine Visit** - This reason is the **`default`** for every exposed infant and is replaced with the other two reasons listed
**NB:** *An exposed infant can have only one reason listed on the appointments list. Milestone HIV test and Dry Blood Sample Positive replace Routine Visit; and Dry Blood Sample Positive replaces Milestone HIV test*

The **Child Contact**, **Sexual Contact** or **Person Who was Reactive at Community** will be on the tracing list for 1 reason:

 1. **HIV Test** - This reason will be added if the Child Contact, Sexual Contact or Person Who was Reactive at Community have a *date of clinic appointment entered* through the **`Tracing Outcomes`** and **`HIV Test`** and **`Next Appointment questionnaires`**

## FHIR Resource to track addition of patient to Appointment List
The following aspects of addition of patients into the Appointment List for Malawi Core will need to be represented in FHIR:

 - Representing addition to the Appointment List
 - Representing the appointment date
 - Representing the reason(s) for the Appointment
 - Representing fulfillment of the appointment
 
 ### [Appointment Resource](https://www.hl7.org/fhir/appointment.html)
 
 The proposed resource to track addition to the Appointment List is the [**Appointment Resource**](https://www.hl7.org/fhir/appointment.html).
 The table below outlines the adopted elements of the Appointment Resource:
 
| Element | Description/Example |
| ------ | ------ |
| `resourceType`  | Appointment|
| `id` | logical identifier for the resource (UUID) |
| `status` | One of the [AppointmentStatus](https://www.hl7.org/fhir/valueset-appointmentstatus.html) as outlined in the Appointment Status Section below |
| `serviceCategory` | [CodeableConcept](https://www.hl7.org/fhir/datatypes.html#CodeableConcept) to represent the categorization of the service to be performed during the appointment. We will utilize the [ActCode](https://build.fhir.org/ig/HL7/UTG/CodeSystem-v3-ActCode.html) code represending HIV/AIDs Program. The value will be `[{"coding": [{"system": "http://terminology.hl7.org/CodeSystem/v3-ActCode","code": "HIVAIDS","display": "HIV-AIDS program"}],"text" : "HIV-AIDS program"}]` |
| `appointmentType` | [CodeableConcept](https://www.hl7.org/fhir/datatypes.html#CodeableConcept) to represent the style of the appointment. The value will be `{"coding": [{"system": "http://terminology.hl7.org/CodeSystem/v2-0276","code": "ROUTINE","ROUTINE": "Routine appointment"}],"text" : "Routine appointment"}` |
| `reasonCode` | [CodeableConcept](https://www.hl7.org/fhir/datatypes.html#CodeableConcept) array to represent one or more reasons for the appointment as outlined in the Appointment Reasons section below. example value to represent `ICT` and `Refill` reasons: `[{"coding": [{"system": "https://d-tree.org","code": "ICT","display": "Index Case Testing"}]},{"coding": [{"system": "https://d-tree.org","code": "Refill","display": "Refill"}]}]` |
| `start` | [instant](https://www.hl7.org/fhir/datatypes.html#instant) in the format `YYYY-MM-DDThh:mm:ss.sss+zz:zz` indicating when the Appointment is to take place. In Malawi Core, the date of the appointment is considered and not a specific time. The instance can thus be zero filled after the date. Example value is `2023-01-01T00:00:00.000Z` |
| `created` | The [dateTime](https://www.hl7.org/fhir/datatypes.html#dateTime) that the appointment was initially created. |
| `participant` | [BackboneElement](https://www.hl7.org/fhir/backboneelement.html) containing the reference to both the patient and the practitioner. Example value is. `[{ "type": [ { "coding": [ { "system": "http://terminology.hl7.org/CodeSystem/v3-ParticipationType", "code": "SBJ", "display": "subject" } ], "text": "Subject" } ], "actor": { "reference": "Patient/e5cee65b-ef98-47ae-aaac-a393014981bd" }, "required": "required", "status": "accepted" }, { "type": [ { "coding": [ { "system": "http://terminology.hl7.org/CodeSystem/v3-ParticipationType", "code": "PRF", "display": "performer" } ], "text": "Performer" } ], "actor": { "reference": "Practitioner/a2efa19f-e87b-4915-8ba2-18096b57e33c" }, "required": "required", "status": "accepted" } ]` |

### Appointment Status
The following are the [AppointmentStatus](https://www.hl7.org/fhir/valueset-appointmentstatus.html) codes 
that we will utilize to represent Appointments in Malawi Core from the possible value set defined by the FHIR project. 
| Code | Definition |
| ------ | ------ |
| `booked` | The default code when extracting the Malawi Core Appointments denoting that it is confirmed for the specified date. |
| `fulfilled` | The Appointment Resource is updated with this code when the **Finish Visit Questionnaire** is extracted before or on the specified date of the appointment to denote that the appointment is fulfilled. *NB*: The Appointment Resource is not updated with this code if the Finish Visit Questionnaire is extracted after the date of the appointment. It will remain in booked status. This way, if anyone looks at past dates, the people listed there will be those that missed their appointment only  |

### Appointment Reason Codes 
The following are the custom codes that will be utilized to represent the different Appointment Reasons in the `reasonCode` element of the Appointment Resource.
| Appointment Reason | system | code | display |  text |
| ------ | ------ | ------ | ------ | ------ |
| Cervical Cancer Screening | `https://d-tree.org` | `VIA` | `Cervical Cancer Screening` | `Cervical Cancer Screening` |
| DBS Positive | `https://d-tree.org` | `DBS Pos` | `DBS Positive` | `DBS Positive` |
| HIV Test | `https://d-tree.org` | `HIV Test` | `HIV Test` | `HIV Test` |
| Index Case Testing | `https://d-tree.org` | `ICT` | `Index Case Testing` | `Index Case Testing` |
| Linkage | `https://d-tree.org` | `Linkage` | `Linkage` | `Linkage` |
| Milestone HIV Test | `https://d-tree.org` | `Milestone` | `Milestone HIV Test` | `Milestone HIV Test` |
| Refill | `https://d-tree.org` | `Refill` | `Refill` | `Refill` |
| Routine Visit | `https://d-tree.org` | `Routine` | `Routine Visit` | `Routine Visit` |
| Viral Load Collection | `https://d-tree.org` | `VL` | `Viral Load Collection` | `Viral Load Collection` |
| Welcome Service | `https://d-tree.org` | `Welcome` | `Welcome Service` | `Welcome Service` |
| Welcome Service Follow Up | `https://d-tree.org` | `Welcome Service Follow Up` | `Welcome Service Follow Up` | `Welcome Follow Up` |

**Comment by roywanyaga** [comment]
### Example Appointment Reason JSON

```
{
    "resourceType": "Appointment",
    "status": "booked",
    "serviceCategory": [
      {
        "coding": [
          {
            "system": "http://terminology.hl7.org/CodeSystem/v3-ActCode",
            "code": "HIVAIDS",
            "display": "HIV-AIDS program"
          }
        ],
        "text": "HIV-AIDS program"
      }
    ],
    "appointmentType": {
      "coding": [
        {
          "system": "http://terminology.hl7.org/CodeSystem/v2-0276",
          "code": "ROUTINE",
          "ROUTINE": "Routine appointment"
        }
      ],
      "text": "Routine appointment"
    },
    "reasonCode": [
      {
        "coding": [
          {
            "system": "https://d-tree.org",
            "code": "ICT",
            "display": "Index Case Testing"
          }
        ]
      },
      {
        "coding": [
          {
            "system": "https://d-tree.org",
            "code": "Refill",
            "display": "Refill"
          }
        ]
      }
    ],
    "start": "2023-01-01T00:00:00.000Z",
    "created": "2022-12-05T12:12:12.000Z",
    "participant": [
      {
        "type": [
          {
            "coding": [
              {
                "system": "http://terminology.hl7.org/CodeSystem/v3-ParticipationType",
                "code": "SBJ",
                "display": "subject"
              }
            ],
            "text": "Subject"
          }
        ],
        "actor": {
          "reference": "Patient/e5cee65b-ef98-47ae-aaac-a393014981bd"
        },
        "required": "required",
        "status": "accepted"
      }
    ]
  }

```

**Comment by f-odhiambo** [comment]
**Some comments** 

1. The client will appear twice on the appointment list if they have been added to the appointment list through both of the questionnaires; one for each. => _I would propose we have a patient with 2 listed appointments instead. Why should we have duplicates and how will this be programmatically done?_
2. The ART Clients will be on the Appointments List for 7 reasons: => _Does a FHIRPath expression rule fulfil the requirements here?_ 
3. Entering a new date of the next appointment in the Finish Visit questionnaire before or on the day of the client's appointment is due creates a new listing for the client in which they no longer appear on the date for ‘today’ or the date they were supposed to but a new date in the future = > _is this an update to the previously created appointment?_  
4. Entering a new date of appointment in the Finish Visit questionnaire after the day the client's appointment was due (i.e. the client missed their appointment, creates a new listing for the client in the future but leaves the missed appointment. This way, if anyone looks at past dates, the people listed there will be those that missed their appointment only. => _Do we need to update the status of the previously skipped appointment while we create a new one ?_
5. Refill – This reason is the default reason set for every ART client and is replaced with the other 6 listed reasons => _Does this mean reason 1,2,4,5,6,7?_
6. I think will need a [status](https://fhir-ru.github.io/appointment-definitions.html#Appointment.status) for the appointment
